### ISSUE TITLE
DataProviderIndexRange doesn't trigger OnSizeChanged callback when SetSize is called

### ISSUE THREAD

**Issue by Meorawr** [Stanzilla/WoWUIBugs#157]
The DataProviderIndexRange mixin doesn't trigger its OnSizeChanged event if [the SetSize method](https://github.com/Gethe/wow-ui-source/blob/fb934dc6558497b2a494898a33e0587a0e6bfe3d/SharedXML/DataProvider.lua#L252-L254) is called to change the contents of the data provider. This is inconsistent with [the Flush method](https://github.com/Gethe/wow-ui-source/blob/fb934dc6558497b2a494898a33e0587a0e6bfe3d/SharedXML/DataProvider.lua#L279-L283) on the mixin, which modifies the size and does trigger the event.

The following snippet and attached video demonstrates the issue; a scrollbox will be placed in the middle of the screen connected to an index-range data provider with an initial size of 10. A numeric spinner is provided above the scrollbox which changes the size of the data provider.

```lua
local DataProvider = CreateDataProviderIndexRange(10);

local ScrollView = CreateScrollBoxListLinearView();
ScrollView:SetDataProvider(DataProvider);
ScrollView:SetElementExtent(28);
ScrollView:SetElementInitializer("Button", "SharedButtonSmallTemplate", function(button, index)
    button:SetText(index);
end);

local ScrollBox = CreateFrame("Frame", nil, UIParent, "WowScrollBoxList");
ScrollBox:SetPoint("CENTER");
ScrollBox:SetSize(400, 400);
ScrollBox:Init(ScrollView);

local SpinBox = CreateFrame("EditBox", nil, UIParent, "NumericInputSpinnerTemplate");
SpinBox:SetPoint("BOTTOM", ScrollBox, "TOP");
SpinBox:SetMinMaxValues(0, math.huge);
SpinBox:SetValue(DataProvider:GetSize());
SpinBox:SetOnValueChangedCallback(function(_, value)
    DataProvider:SetSize(value);
end);
```

As shown in the attached video below, clicking the increase/decrease buttons has no impact on the visual contents of the scrollbox.

https://user-images.githubusercontent.com/287102/128724123-851cc40d-f252-4795-b32e-99d8efc04799.mp4

If the value change callback on the numeric spinner is changed to manually trigger the OnSizeChanged callback, the demo works as expected.

```lua
SpinBox:SetOnValueChangedCallback(function(_, value)
    DataProvider:SetSize(value);
    DataProvider:TriggerEvent("OnSizeChanged", false);
end);
```

https://user-images.githubusercontent.com/287102/128724135-82fbd313-3895-421a-a1eb-5b9be6bf878a.mp4

**Comment by Meorawr** [comment]
Fixed in 2.5.2.39801 for BCC; will wait until fixed on retail before closing.

**Comment by Meorawr** [comment]
Fixed in 1.14.0.39958, 2.5.2.40011, and 9.1.5.39977.
